server/src/main/java/org/apache/druid/metadata/SQLMetadataConnector.java

```diff
@@ -636,7 +638,21 @@ protected BasicDataSource getDatasource()
   {
     MetadataStorageConnectorConfig connectorConfig = getConfig();
 
-    BasicDataSource dataSource = new BasicDataSource();
+    BasicDataSource dataSource = null;
+
+    try {
+      Properties dbcpProperties = connectorConfig.getProperties();
+      if (dbcpProperties != null) {
+        dataSource = BasicDataSourceFactory.createDataSource(dbcpProperties);
+      } else {
+        dataSource = new BasicDataSource();
+      }
+    }
+    catch (Exception e) {
+      log.error(e, "Failed to create data source using dbcp properties file");
+      throw new RuntimeException(e);
+    }
+
     dataSource.setUsername(connectorConfig.getUser());
```

Comment:
It looks like if the user provides the username, password and connectionURL in the dbcp config file as well in the runtime properties, the values from dbcp config file would be ignored. Is that the expected behavior?

Reply:
Yes, not changing existing behavior.  Just adding ability to set all other properties which are not directly settable.